test/test_foreach.py

```diff
@@ -153,54 +120,881 @@ def _test_bin_op_list_alpha(self, device, dtype, foreach_op, foreach_op_, torch_
     #
     # Unary ops
     #
-    @dtypes(*(torch.testing.floating_and_complex_types_and(torch.bfloat16, torch.half)))
-    def test_unary_ops(self, device, dtype):
-        for fe_op, fe_op_, torch_op, support_bfloat16, support_complex in self.unary_ops:
-            for N in N_values:
-                tensors1 = self._get_test_data(device, dtype, N)
-                # Mimics cuda kernel dtype flow.  With fp16/bf16 input, runs in fp32 and casts output back to fp16/bf16.
-                control_dtype = torch.float32 if (self.device_type == 'cuda' and
-                                                  (dtype is torch.float16 or dtype is torch.bfloat16)) else dtype
+    @dtypes(*torch.testing.get_all_dtypes())
+    def test_neg(self, device, dtype):
+        for N in N_values:
+            tensors = self._get_test_data(device, dtype, N)
```

Comment:
We should really consider having OpInfo based tests because I think they will probably eliminate the need for a lot of this logic. I'm not sure if OpInfo covers all the operators that you have here, or how difficult it is to add a new OpInfo into the op database.

If we want to go ahead with the testing without using OpInfo, then I prefer the old testing more than the new testing. The older testing is a lot more concise, but has the following downsides:
1. the branches are a little difficult to follow and end up special casing certain operators. I.e., there isn't a clean "mechanism vs policy" split.
2. Everything runs in the same test case so it's hard to tell which operator is erroring out.

We can fix both of the above problems. For problem 1, I think we just need to add more to the unary_ops table:
```
 foreach_op, foreach_op_, torch_op, cpu_half, cuda_half, cpu_bf16, cuda_bf16, complex64/128, bool
```
this would separate the mechanism (the testing logic) from the policy (the operator table).

For the second problem, we could autogenerate the individual tests by iterating through the table and assigning to the test class. I am not sure how well this works with all the testing decorators though.